c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -3625,6 +3630,388 @@ bool response_started(JSObject *self) {
+  }
+
+  // 4. If data contains a code point that is not one of
+  //    U+002B (+)
+  //    U+002F (/)
+  //    ASCII alphanumeric
+  // then return failure.
+
+  // Step 4 is handled within the calls below to
+  // base64Decode4to3, base64Decode3to2, and base64Decode2to1
+
+  // 5. Let output be an empty byte sequence.
+  std::string output = "";
+  output.reserve(data_view.length() / 3);
+
+  // 6. Let buffer be an empty buffer that can have bits appended to it.
+
+  // Step 6 is handled within the calls below to
+  // base64Decode4to3, base64Decode3to2, and base64Decode2to1
+
+  // 7. Let position be a position variable for data, initially pointing at the start of data.
+
+  // We don't use a position variable, instead we remove_prefix from the `data` each time we have
+  // dealt with some characters.
+
+  while (data_view.length() >= 4) {
+    MOZ_TRY(base64Decode4to3(data_view, output));
+    data_view.remove_prefix(4);
+  }
+
+  switch (data_view.length()) {
+  case 3: {
+    MOZ_TRY(base64Decode3to2(data_view, output));
+    break;
+  }
+  case 2: {
+    MOZ_TRY(base64Decode2to1(data_view, output));
+    break;
+  }
+  case 1:
+    return JS::Result<std::string>(JS::Error());
+  case 0:
+    break;
+  default:
+    MOZ_CRASH("Too many characters leftover");
+  }
+
+  return output;
+}
+
+// https://html.spec.whatwg.org/multipage/webappapis.html#dom-atob
+bool atob(JSContext *cx, unsigned argc, Value *vp) {
+  CallArgs args = CallArgsFromVp(argc, vp);
+  if (!args.requireAtLeast(cx, "atob", 1)) {
+    return false;
+  }
+  auto dataResult = ConvertJSValueToByteString(cx, args.get(0));
+  if (dataResult.isErr()) {
+    return false;
+  }
+  auto data = dataResult.unwrap();
+
+  // 1. Let decodedData be the result of running forgiving-base64 decode on data.
+  auto decoded_result = forgivingBase64Decode(data);
+  // 2. If decodedData is failure, then throw an "InvalidCharacterError" DOMException.
+  if (decoded_result.isErr()) {
+    JS_ReportErrorNumberUTF8(cx, GetErrorMessage, nullptr, JSMSG_INVALID_CHARACTER_ERROR);
+    return false;
+  }
+  auto decoded = decoded_result.unwrap();
+  RootedString decodedData(cx, JS_NewStringCopyN(cx, decoded.c_str(), decoded.length()));
+  if (!decodedData) {
+    return false;
+  }
+
+  // 3. Return decodedData.
+  args.rval().setString(decodedData);
+  return true;
+}
+
+const char base[65] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+                      "abcdefghijklmnopqrstuvwxyz"
+                      "0123456789+/";
+
+inline uint8_t CharTo8Bit(char character) { return uint8_t(character); }
+inline void base64Encode3to4(std::string_view data, std::string &output) {
+  uint32_t b32 = 0;
+  int i, j = 18;
+
+  for (i = 0; i < 3; ++i) {
+    b32 <<= 8;
+    b32 |= CharTo8Bit(data[i]);
+  }
+
+  for (i = 0; i < 4; ++i) {
+    output += base[(uint32_t)((b32 >> j) & 0x3F)];
+    j -= 6;
+  }
+}
+
+inline void base64Encode2to4(std::string_view data, std::string &output) {
+  uint8_t src0 = CharTo8Bit(data[0]);
```

Comment:
I like `static_cast` with an `auto` result for cases like these, what do you think?
```suggestion
  auto src0 = static_cast<uint8_t>(data[0]);
```

Reply:
I like it, I applied this suggestion as part of [`ae81dca` (#227)](https://github.com/fastly/js-compute-runtime/pull/227/commits/ae81dca453df75f519a3af68046348497954ab0e)